### ISSUE TITLE
Clean implementation of AptaMCTS

### ISSUE THREAD

**Issue by NennoMP** [gc-os-ai/pyaptamer#21]
_UPDATED: 07-04-2025_

Add a clean implementation of AptaMCTS, an aptamer recommendation pipeline which combines RandomForest and Monthe Carlo Tree Search (MCTS). These components are the model/classifier and the string optimization algorithm, respectively.

**Todo.** Add a clean, simplified Python implementation of AptaMCTS's model/classifier and preprocessing pipeline to the repository. The MCTS algorithm is already implemented and can easily be used by model with a compatible interface (e.g., see AptaNet and AptaTrans).

High-level steps could be the following:
- Look at the authors' code, specifically the `paper_version` folder, for insights into the original implementation.
- Refactor and add to the repository the model/classifier (in `classifier.py`) and any other data pre-processing logic characteristic of AptaMCTS.
- Implement corresponding test batteries.

AptaMCTS is mostly a RandomForest classifier and leverages `sklearn`. I would expect a very similar interface to that of AptaNet, with the following classes:
- A pipeline module, `aptamcts._pipeline.py`, for the overall pipeline logic. 
- If needed, a model module, `aptamcts._feature_classifier.py`, implementing a classifier and regressor version. This module could also contain specific data preprocessing steps, if any.
- An experiment module, `experiments._aptamer_aptamcts.py`, so that AptaMCTS classifier may be combined with the current implementation of MCTS.

In the end, we want a clean AptaMCTS implementation, easily usable for training and predicting, and easily usable in combination with the current implementation of MCTS for recommending aptamer candidates.

**Code:** [https://github.com/leekh7411/Apta-MCTS](https://github.com/leekh7411/Apta-MCTS).
**Paper:** Lee, Gwangho, et al. "Predicting aptamer sequences that interact with target proteins using an aptamer-protein interaction classifier and a Monte Carlo tree search approach." PloS one 16.6 (2021): e0253760.

**Comment by satvshr** [comment]
Can this be closed given MCTS is already in the repo @fkiraly ?

**Comment by fkiraly** [comment]
No, becaues MCTS is not the same as AptaMCTS

**Comment by siddharth7113** [comment]
@NennoMP  Can you add a summarised/updated instrcutions for new contributors to pick this issue , and if possible along with reference to current code in regards to style, formatting and api design for this issue.

**Comment by aditi-dsi** [comment]
@siddharth7113 I am looking at this issue and will update the progress here.

**Comment by aditi-dsi** [comment]
Hi @fkiraly and @NennoMP, I took some time to go through the referenced paper and repos, reviewed the current implementation of MCTS in the repository along with the back-n-forth discussion above. I want to confirm my understanding of the discussion regarding the scope of this work. Please correct me if I'm wrong:

It seems the current MCTS implementation in the repo is specifically the string-optimization algorithm (which is apparently called AptaMCTS). My understanding is that the expected outcome (i.e., to adhere to separation of concerns) is:
- to make MCTS a generic abstract class (similar to `BaseOptimizer` in `hyperactive`), 
- and then create an AptaMCTS class (or refactor the existing code) that imports and inherits from `BaseMCTS` to handle specifically the sequence reconstruction for aptamer generation task.

Per Franz's suggestion, I tried to draw a rough usage vignette (will write docstrings after confirming the idea). I split them to confirm the two different usage types:

 - Internal API (for building custom optimization pipelines like AptaMCTS based on MCTS)

```python
from pyaptamer.optimizers import BaseMCTS

# Developers can subclass the base class to define custom logic
class CustomMCTS(BaseMCTS):
    def _reconstruct(self, sequence):
        # some custom sequence handling logic
        return sequence

custom_optimizer = CustomMCTS(states=["A", "B"], depth=10)
best_sequence = custom_optimizer.run(objective_function=custom_score_function)
```

- Public API (out-of-the-box usage for domain tasks):

```python
from pyaptamer.mcts import AptaMCTS

# Inherits from BaseMCTS internally, hiding the state config
apta_optimizer = AptaMCTS(depth=20, n_iterations=1000, experiment=experiment)
candidate = apta_optimizer.run()
```

Does this scope look correct, or am I missing something?

**Comment by aditi-dsi** [comment]
> Therefore, it should fit an API that is a unified API for generic optimization algorithms - specifically, that of `hyperactive` `BaseOptimizer`

Just a quick follow-up regarding this `hyperactive` reference. Should the abstract `BaseMCTS` class be built inside `pyaptamer` itself, or is the actual goal to implement it directly in the `hyperactive` repository and import it here?

**Comment by NennoMP** [comment]
@aditi-dsi The current `MCTS` class implements the Monte Carlo Tree Search algorithm used inside AptaMCTS (not yet implemented) and also AptaTrans (already implemented). AptaMCTS combines MCTS with a specific model, used inside MCTS as a scoring function to evaluate most promising candidates. I believe the specific model used was a RandomForest algorithm, see the original authors' code. 

Thus, what's currently missing to have AptaMCTS fully implemented is this: a pipeline class (similar to that of AptaTrans or AptaNet) implementing the model logic and/or data preprocessing. MCTS should not be changed as it is supposed to work with any model with a compatible interface.

I agree that the previous description was not clear and updated it.

**Comment by aditi-dsi** [comment]
@NennoMP thanks for updating the details, it's very helpful. I have gone through them and will start working on this now. I'll open a draft PR shortly so you can track the progress.

**Comment by aditi-dsi** [comment]
@NennoMP sorry for the delay, got a bit occupied due to uni exams. 

As I was reviewing the `paper_version` code, I have a quick question on the encoding method before I implement it.  I noticed a discrepancy between the original paper and the authors' code. The paper concludes that TPC + PseKNC is the optimal encoding combination but the authors' provided code (`preprocess.py`) only implements iCTF. 

So, should I stick to refactoring the iCTF logic from their code, or would you prefer I implement TPC + PseKNC as mentioned in the original paper?

**Comment by NennoMP** [comment]
> [@NennoMP](https://github.com/NennoMP) sorry for the delay, got a bit occupied due to uni exams.
> 
> As I was reviewing the `paper_version` code, I have a quick question on the encoding method before I implement it. I noticed a discrepancy between the original paper and the authors' code. The paper concludes that TPC + PseKNC is the optimal encoding combination but the authors' provided code (`preprocess.py`) only implements iCTF.
> 
> So, should I stick to refactoring the iCTF logic from their code, or would you prefer I implement TPC + PseKNC as mentioned in the original paper?

I would stick to what's in the code.

**Comment by DashratRajpurohit** [comment]
Hi @NennoMP! I saw this issue for the` AptaMCTS `cleanup and wanted to contribute something...

I've put together a working implementation locally. I focused on making it 'Pyaptamer-style' by building a proper `K-mer encoder` and a **Scikit-learn pipeline** wrapper. This makes it really easy to swap models or k-mer sizes during testing.

I also added the **AptamerEval **wrapper so we can actually run it through the benchmarking system right away. I've been careful with the type hints and docstrings so it's as clean as possible.

Let me know if you want me to open a PR—I'd love to get your feedback on the structure! as a student it helps me in my journey``

**Comment by DashratRajpurohit** [comment]
I've finished the unit tests for my implementation. All modules (K-mer encoding and the full sklearn-style pipeline) are passing successfully with 100% coverage on core logic. Ready to PR whenever you're free to review.......

**Comment by kris70lesgo** [comment]
@NennoMP @siddharth7113 @satvshr 
I’ve opened this pr implementing a minimal AptaMCTS pipeline following the AptaTrans-style design.
I’d appreciate any feedback on the architecture, API consistency, and MCTS integration when you get a chance. Thanks!

**Comment by Whatsonyourmind** [comment]
Two design decisions worth pinning before the implementation since they're hard to reverse later:

**1. How the classifier's outputs feed back into MCTS expansion.**

If the RandomForest exposes calibrated `predict_proba`, you have two architectural choices:

- **Pure rollout**: classifier scores at leaf nodes, UCB1 backup. Simple, works well when the classifier is accurate but slow (it's only called at terminals).
- **PUCT priors** (AlphaGo Zero style): classifier scores feed in as priors `P(a|s)` at every expansion. Selection becomes `Q(s,a) + c · P(a|s) · √N(s) / (1 + N(s,a))`. Better sample efficiency at the cost of per-step classifier calls.

For aptamer sequence search the per-step branching factor is small (4 nucleotides + length decisions), so PUCT priors are tractable and probably worth it — Lee et al. use pure rollout, but a clean rewrite has room to improve. Worth deciding upfront because the `classifier.py` interface differs between them: `predict_proba(state)` vs `score_terminal(state)`.

**2. Calibration of classifier outputs before MCTS uses them.**

Random Forest's `predict_proba` is famously badly calibrated — overconfident at the extremes, underconfident in the middle. If the same classifier is used both for terminal scoring AND for tree priors, miscalibration compounds: an overconfident leaf score biases all paths through that subtree. Two cheap fixes:

- **Isotonic regression** (Niculescu-Mizil & Caruana 2005) on a held-out validation split. ~50 lines, fixes the worst of the miscalibration; `sklearn.isotonic.IsotonicRegression` is the drop-in.
- **Beta calibration** (Kull, Silva Filho, Flach 2017) — closed-form, parametric, works better when calibration-set sample size is small (relevant here since aptamer datasets are usually <10K).

Recommend running calibration in `_pipeline.py` rather than `_feature_classifier.py` so multiple classifier types (AptaNet, AptaTrans, AptaMCTS) can share the calibration layer with one implementation.

**3. One detail orthogonal to either choice — the exploration constant `c`.**

The standard `c = √2` is for [-1, 1] reward range. If the classifier outputs probabilities in [0, 1], the optimal `c` shifts to roughly `√2 / 2 ≈ 0.707`. Worth surfacing as a parameter in the experiment module so users tune it explicitly rather than inheriting it implicitly from a non-matching default.

Happy to sketch the `_pipeline.py` interface signatures or a `Calibrator` class once the classifier API is settled.
